### PR TITLE
WIP: Expose raw RRD data to LibreNMS app

### PR DESCRIPTION
**Pull request by haydenseitz**
This will most likely be a small PR, but hopefully the first of many to allow querying the data stored in RRD. The (over-ambitious) plan is to expose RRD data to LibreNMS internal libraries, which will allow for future improvements, such as a UI graphing replacement and to allow graphing from another supported data source (graphite, influx, etc).

Please share any suggestions, and if this conflicts with other future project plans. I've made a few assumptions on best-practice, such as using the `$json_enabled` variable in graph include files, like `uptime.inc.php`. This seems to be a decent way to only enable JSON results as the include files are updated to support them.

I'm on the discord at `@hayden#8083`

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by willhseitz** [comment]
Here's an example of an API call using `output=json`:

curl:
`http://localhost:8000/api/v0/devices/localhost/device_uptime?from=-5m&output=json`

results:
```
{
    "status": "ok",
    "dataset": {
        "about": "RRDtool graph JSON output",
        "meta": {
            "start": 1599782280,
            "end": 1599782400,
            "step": 60,
            "legend": [
                "Uptime"
            ]
        },
        "data": [
            [
                "1599782280",
                220.13000891
            ],
            [
                "1599782340",
                220.13073511
            ]
        ]
    },
    "count": 2
}
```

The idea is then to take the raw data and use to graph, etc

File: includes/html/api_functions.inc.php [hunk 1]
```diff
@@ -97,6 +97,17 @@ function api_get_graph(array $vars)
         return api_success(['image' => $base64_output, 'content-type' => get_image_type()], 'image');
     }
 
+    if ($vars['output'] === 'json') {
+        $json = json_decode($image);
+        if ($json !== null) {
+            return api_success($json, 'dataset', null, 200, count($json->data));
+        }
+        if ($json_enabled) {
```
Comment (Jellyfrog) [inline includes/html/api_functions.inc.php:105]: The variable ``$json_enabled`` seems to be never defined.

Comment (willhseitz) [reply includes/html/api_functions.inc.php:105]: For now, that variable is only set in `includes/html/graph/devices/uptime.inc.php`. The idea is when each graph type gets updated to support the JSON output, that variable will be set.
A better way might be to use a function like `isJSONenabled($graph_type)` that checks a list of JSON supported graph types

**Comment by murrant** [comment]
I like your thoughts here.  Keep in mind, graphing is one of the oldest components in LibreNMS right now.   It is probably going to be easier and have a better outcome to just start from scratch.  I have updated the "Datastores" to something more modern, but only for storing data, not retrieving it.  In an ideal system, we can even have all graphs be live updating.

**Comment by willhseitz** [comment]
> graphing is one of the oldest components in LibreNMS right now

It shows 😆

I have two minds about how to approach graphing:
1. Make incremental changes to move from the current setup of RRD graphing with RRD datastore, to something like Chart.js graphing with any supported datastore
2. Start a separate branch or fork and do it the "start from scratch" way

Benefit of option 1 is it could become a completely back-end agnostic platform, but would of course be a lot more work

**Comment by willhseitz** [comment]
Here's another example for the `device_bits` graph, and making a request directly to graph.php (not the API):

Request:
`http://localhost:8000/graph.php?type=device_bits&device=1&from=-10m&to=now&output=json`

Results:
```
{ "about": "RRDtool graph JSON output",
  "meta": {
    "start": 1599862440,
    "end": 1599863040,
    "step": 60,
    "legend": [
      "ipsec0 In",
      "ipsec1 In",
      "ipsec2 In",
      "ipsec3 In",
      "ixp0 In",
      "ixp1 In",
      "ipsec0 Out",
      "ipsec1 Out",
      "ipsec2 Out",
      "ipsec3 Out",
      "ixp0 Out",
      "ixp1 Out"
          ]
     },
  "data": [
    [ "1599862440",null, null, null, null, 5.1723000814e+03, 4.2337297675e+02, null, null, null, null, -4.7830857282e+03, 0.0000000000e+00 ],
    [ "1599862500",null, null, null, null, 6.9410232883e+03, 5.7660608311e+02, null, null, null, null, -6.0341558761e+03, 0.0000000000e+00 ],
    [ "1599862560",null, null, null, null, 4.6737124127e+03, 4.1975392493e+02, null, null, null, null, -3.7757985703e+03, 0.0000000000e+00 ],
    [ "1599862620",null, null, null, null, 7.4888189243e+03, 5.6783372395e+02, null, null, null, null, -7.1293224796e+03, 0.0000000000e+00 ],
    [ "1599862680",null, null, null, null, 5.1291558687e+03, 4.1640945466e+02, null, null, null, null, -5.0243742033e+03, 0.0000000000e+00 ],
    [ "1599862740",null, null, null, null, 7.0045179432e+03, 5.5852862999e+02, null, null, null, null, -6.0095441038e+03, 0.0000000000e+00 ],
    [ "1599862800",null, null, null, null, 4.0823835209e+03, 3.4449367800e+02, null, null, null, null, -3.0573671402e+03, 0.0000000000e+00 ],
    [ "1599862860",null, null, null, null, 7.1130231423e+03, 6.3277674544e+02, null, null, null, null, -5.6154579929e+03, 0.0000000000e+00 ],
    [ "1599862920",null, null, null, null, 6.1956070293e+03, 6.8404211171e+02, null, null, null, null, -4.3318563697e+03, 0.0000000000e+00 ],
    [ "1599862980",null, null, null, null, null, null, null, null, null, null, null, null ]
  ]
}
```

The data is pretty clunky. The first element in each row is the epoch, followed by each data point listed in the legend. We could always json_decode, manipulate the data however we like, then json_encode for output. That may help if the goal is to allow JSON results from any datastore backend

**Comment by murrant** [comment]
Closing for now, this needs a much more robust approach.

**Pull request closed without merging**